### PR TITLE
Added controller for running ansible playbooks

### PR DESCRIPTION
**Pull request by eshulman2**
This addition should be flexible enough to run
ansible playbooks for testing popuses.
it is dependent on this patch
https://github.com/openstack-k8s-operators/tcib/pull/182

### CODE REVIEW

**Comment by lpiwowar** [comment]
Hi @eshulman2, thanks for the PR! This is going to be helpful once we merge this!

I will do a deeper review a little bit later but in the meantime there are some failures that need to be fixed [1]:
```
 # github.com/openstack-k8s-operators/test-operator/controllers
2024-05-21 07:39:12.311503 | controller |     controllers/trex_performance_controller.go:21:2: "fmt" imported and not used
2024-05-21 07:39:12.311509 | controller |     controllers/trex_performance_controller.go:168:3: not enough arguments in call to r.EnsureLogsPVCExists
2024-05-21 07:39:12.311516 | controller |             have (context.Context, *"github.com/openstack-k8s-operators/test-operator/api/v1beta1".TrexPerformance, *helper.Helper, map[string]string, string)
2024-05-21 07:39:12.311522 | controller |             want (context.Context, client.Object, *helper.Helper, map[string]string, string, bool)
2024-05-21 07:39:12.311533 | controller |     controllers/trex_performance_controller.go:273:24: instance.Spec.TrexPerformanceSpec undefined (type "github.com/openstack-k8s-operators/test-operator/api/v1beta1".TrexPerformanceSpec has no field or method TrexPerformanceSpec)
2024-05-21 07:39:12.311542 | controller |     controllers/trex_performance_controller.go:276:51: instance.Spec.Workflow[workflowStepNum].TrexPerformanceSpec undefined (type "github.com/openstack-k8s-operators/test-operator/api/v1beta1".TrexPerformanceWorkflowSpec has no field or method TrexPerformanceSpec)
2024-05-21 07:39:12.311550 | controller |     controllers/trex_performance_controller.go:282:58: wtRun.Serial undefined (type "github.com/openstack-k8s-operators/test-operator/api/v1beta1".TrexPerformanceWorkflowSpec has no field or method Serial)
2024-05-21 07:39:12.311557 | controller |     controllers/trex_performance_controller.go:290:2: undefined: value
2024-05-21 07:39:12.311563 | controller |     controllers/trex_performance_controller.go:291:9: undefined: value
2024-05-21 07:39:12.311570 | controller |     controllers/trex_performance_controller.go:292:40: undefined: value
2024-05-21 07:39:12.311576 | controller |     controllers/trex_performance_controller.go:297:6: mergeWithWorkflow redeclared in this block
2024-05-21 07:39:12.311583 | controller |             controllers/tempest_controller.go:471:6: other declaration of mergeWithWorkflow
2024-05-21 07:39:12.311589 | controller |     controllers/trex_performance_controller.go:317:62: undefined: tempest
2024-05-21 07:39:12.311604 | controller |     controllers/trex_performance_controller.go:21:2: too many errors
```
I recommend at least running the test-operator with `make install run` before pushing the change.

[1] https://logserver.rdoproject.org/97/97/77d2a159a959cc584d3f36d7305e3202c1e412f0/github-check/openstack-k8s-operators-content-provider/d166321/job-output.txt

File: api/v1beta1/trex_performance_types.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="local-storage"
+	// StorageClass used to create PVCs that store the logs
+	StorageClass string `json:"storageClass"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="dataplane-ansible-ssh-private-key-secret"
+	// ComputeSSHKeySecretName is the name of the k8s secret that contains an ssh key for computes.
+	// The key is mounted to ~/.ssh/id_ecdsa in the tempest pod
```
Comment (lpiwowar) [inline api/v1beta1/trex_performance_types.go:52]: nit: tempest -> test pod (or TrexPerformance pod)

File: api/v1beta1/trex_performance_types.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	ComputesSSHKeySecretName string `json:"computeSSHKeySecretName"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default=""
+	// WorkloadSSHKeySecretName is the name of the k8s secret that contains an ssh key for the performance workload.
+	// The key is mounted to ~/test_keypair.key in the tempest pod
```
Comment (lpiwowar) [inline api/v1beta1/trex_performance_types.go:58]: nit: tempest -> test pod (or TrexPerformance pod)

File: api/v1beta1/trex_performance_types.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="local-storage"
+	// StorageClass used to create PVCs that store the logs
+	StorageClass *string `json:"storageClass"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="dataplane-ansible-ssh-private-key-secret"
+	// ComputeSSHKeySecretName is the name of the k8s secret that contains an ssh key for computes.
+	// The key is mounted to ~/.ssh/id_ecdsa in the tempest pod
```
Comment (lpiwowar) [inline api/v1beta1/trex_performance_types.go:133]: ditto

File: api/v1beta1/trex_performance_types.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+	ComputesSSHKeySecretName *string `json:"ComputeSSHKeySecretName"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default=""
+	// WorkloadSSHKeySecretName is the name of the k8s secret that contains an ssh key for the performance workload.
+	// The key is mounted to ~/test_keypair.key in the tempest pod
```
Comment (lpiwowar) [inline api/v1beta1/trex_performance_types.go:139]: ditto

File: controllers/trex_performance_controller.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+// +kubebuilder:rbac:groups="",resources=pods,verbs=create;delete;get;list;patch;update;watch
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;create;update;watch;patch
+
+// Reconcile - TrexPerformanceReconciler
+func (r *TrexPerformanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, _err error) {
+
+	// How much time should we wait before calling Reconcile loop when there is a failure
+	requeueAfter := time.Second * 60
+
+	// Fetch the Tempest instance
```
Comment (lpiwowar) [inline controllers/trex_performance_controller.go:73]: ditto

File: main.go [hunk 6]
```diff
@@ -104,6 +104,14 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "Tobiko")
 		os.Exit(1)
 	}
+
+	performanceReconciler := &controllers.TrexPerformanceReconciler{}
+	performanceReconciler.Client = mgr.GetClient()
+	performanceReconciler.Scheme = mgr.GetScheme()
+	if err = tempestReconciler.SetupWithManager(mgr); err != nil {
```
Comment (lpiwowar) [inline main.go:111]: tempestReconciler -> performanceReconciler

Does it really work with this code as it is now? If yes, I'm little bit surprised. I'll need to test this manually.

File: pkg/trexperf/volumes.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+			ReadOnly:  true,
+		},
+		{
+			Name:      "config-data",
+			MountPath: "/etc/test_operator",
+			ReadOnly:  false,
+		},
+		{
+			Name:      "test-operator-logs",
+			MountPath: "/var/lib/tempest/external_files",
```
Comment (lpiwowar) [inline pkg/trexperf/volumes.go:159]: tempest -> trexPerformance (or something similar)

ditto

File: pkg/trexperf/volumes.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+		},
+		{
+			Name:      "openstack-config",
+			MountPath: "/etc/openstack/clouds.yaml",
+			SubPath:   "clouds.yaml",
+			ReadOnly:  true,
+		},
+		{
+			Name:      "openstack-config",
+			MountPath: "/var/lib/tempest/.config/openstack/clouds.yaml",
```
Comment (lpiwowar) [inline pkg/trexperf/volumes.go:170]: ditto

File: pkg/trexperf/volumes.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+			ReadOnly:  true,
+			SubPath:   "tls-ca-bundle.pem",
+		}
+
+		volumeMounts = append(volumeMounts, caCertVolumeMount)
+	}
+
+	workloadSSHKeyMount := corev1.VolumeMount{
+		Name:      "workload-ssh-secret",
+		MountPath: "/var/lib/tempest/test_keypair.key",
```
Comment (lpiwowar) [inline pkg/trexperf/volumes.go:204]: ditto

File: api/v1beta1/trex_performance_types.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+	// Hash
+	Hash string `json:"hash,omitempty"`
+}
+*/
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// TrexPerformanceSpec defines the desired state of TrexPerformance
+type TrexPerformanceSpec struct {
```
Comment (lpiwowar) [inline api/v1beta1/trex_performance_types.go:38]: Just a thinking ... Shouldn't we choose other name than TrexPerformance? Maybe `AnsibleTests`?

Is the goal to have generic ansible tests runner, right?:)

File: main.go (hunk 6 above)
Comment (eshulman2) [reply main.go:111]: haven't got to actually testing it due to other priorities

**Comment by eshulman2** [comment]
recheck

File: api/v1beta1/trex_performance_types.go (hunk 10 above)
Comment (lpiwowar) [reply api/v1beta1/trex_performance_types.go:28]: @eshulman2 I'm just wondering. Do we still want to go in the direction of the generic `AnsibleTests` CR or you encountered some issues that do not make it possible?

File: api/v1beta1/trex_performance_types.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+	// +kubebuilder:validation:Required
+	// Name of a workflow step. The step name will be used for example to create
+	// a logs directory.
+	StepName string `json:"stepName"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="local-storage"
```
Comment (lpiwowar) [inline api/v1beta1/trex_performance_types.go:35]: Can you please add the following for each parameter and then run `make bundle`? The command will update `config/manifests/bases/test-operator.clusterserviceversion.yaml`. The updated version of this file should be part of the PR.

```
// +operator-sdk:csv:customresourcedefinitions:type=spec
```

File: api/v1beta1/trex_performance_types.go (hunk 10 above)
Comment (eshulman2) [reply api/v1beta1/trex_performance_types.go]: done

File: config/crd/patches/webhook_in_ansibleTests.yaml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+# The following patch enables a conversion webhook for the CRD
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: ansibleTest.test.openstack.org
```
Comment (lpiwowar) [inline config/crd/patches/webhook_in_ansibleTests.yaml:5]: @eshulman2 I think we have to go through the occurrences of the `ansibleTest` in the PR. And think whether it should be capitalized or not. As an example you can take a look at HorizonTest CR. You can see that for example here [1] the name for HorionTests CR is not capitalized. 

Also, checking whether the test-operator works from locally build test-operator-index would be in this case a good idea I think [2] (question #7).

[1] https://github.com/openstack-k8s-operators/test-operator/blob/b68150cf867b538bc9896b9d2102d3c29ded2c29/config/crd/patches/webhook_in_horizontests.yaml#L5
[2] https://openstack-k8s-operators.github.io/test-operator/FAQ.html

**Review by lpiwowar — APPROVED** [review]
LGTM! :+1: 
I think this is a solid foundation for the AnsibleTest CR. We can always address any minor potential issues down the road.

**Pull request merged**